File: packages/wasmvm/wasmlib/src/wasmtypes/screquestid.rs
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use std::convert::TryInto;

use crate::*;

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

pub const SC_REQUEST_ID_LENGTH: usize = 34;
pub const REQUEST_ID_SEPARATOR: &str = "-";

#[derive(PartialEq, Clone)]
pub struct ScRequestID {
    id: [u8; SC_REQUEST_ID_LENGTH],
}

impl ScRequestID {
    pub fn to_bytes(&self) -> Vec<u8> {
        request_id_to_bytes(self)
    }

    pub fn to_string(&self) -> String {
        request_id_to_string(self)
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

pub fn request_id_decode(dec: &mut WasmDecoder) -> ScRequestID {
    request_id_from_bytes_unchecked(&dec.fixed_bytes(SC_REQUEST_ID_LENGTH))
}

pub fn request_id_encode(enc: &mut WasmEncoder, value: &ScRequestID) {
    enc.fixed_bytes(&value.id, SC_REQUEST_ID_LENGTH);
}

pub fn request_id_from_bytes(buf: &[u8]) -> ScRequestID {
    if buf.len() == 0 {
        return ScRequestID {
            id: [0; SC_REQUEST_ID_LENGTH],
        };
    }
    if buf.len() != SC_REQUEST_ID_LENGTH {
        panic("invalid RequestID length");
    }
    // final uint16 output index must be > ledgerstate.MaxOutputCount
    if buf[SC_REQUEST_ID_LENGTH - 2] > 127 || buf[SC_REQUEST_ID_LENGTH - 1] != 0 {
        panic("invalid RequestID: output index > 127");
    }
    ScRequestID {
        id: buf.try_into().expect("WTF?"),
    }
}

pub fn request_id_to_bytes(value: &ScRequestID) -> Vec<u8> {
    value.id.to_vec()
}

pub fn request_id_from_string(value: &str) -> ScRequestID {
    request_id_from_bytes(&hex_decode(value))
}

pub fn request_id_to_string(value: &ScRequestID) -> String {
    hex_encode(&value.id)
}

fn request_id_from_bytes_unchecked(buf: &[u8]) -> ScRequestID {
    ScRequestID {
        id: buf.try_into().expect("invalid RequestID length"),
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

pub struct ScImmutableRequestID {
    proxy: Proxy,
}

impl ScImmutableRequestID {
    pub fn new(proxy: Proxy) -> ScImmutableRequestID {
        ScImmutableRequestID { proxy }
    }

    pub fn exists(&self) -> bool {
        self.proxy.exists()
    }

    pub fn to_string(&self) -> String {
        request_id_to_string(&self.value())
    }

    pub fn value(&self) -> ScRequestID {
        request_id_from_bytes(&self.proxy.get())
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

// value proxy for mutable ScRequestID in host container
pub struct ScMutableRequestID {
    proxy: Proxy,
}

impl ScMutableRequestID {
    pub fn new(proxy: Proxy) -> ScMutableRequestID {
        ScMutableRequestID { proxy }
    }

    pub fn delete(&self) {
        self.proxy.delete();
    }

    pub fn exists(&self) -> bool {
        self.proxy.exists()
    }

    pub fn set_value(&self, value: &ScRequestID) {
        self.proxy.set(&request_id_to_bytes(&value));
    }

    pub fn to_string(&self) -> String {
        request_id_to_string(&self.value())
    }

    pub fn value(&self) -> ScRequestID {
        request_id_from_bytes(&self.proxy.get())
    }
}

File: packages/wasmvm/wasmlib/as/wasmlib/wasmtypes/screquestid.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import {panic} from "../sandbox";
import {concat, hexDecode, hexEncode, WasmDecoder, WasmEncoder, zeroes} from "./codec";
import {uint16FromBytes, uint16FromString, uint16ToBytes, uint16ToString} from "./scuint16";
import {Proxy} from "./proxy";
import {bytesCompare} from "./scbytes";
import {hashFromBytes, hashToBytes} from "./schash";

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export const ScRequestIDLength = 34;
const RequestIDSeparator = "-"

export class ScRequestID {
    id: Uint8Array = zeroes(ScRequestIDLength);

    public equals(other: ScRequestID): bool {
        return bytesCompare(this.id, other.id) == 0;
    }

    // convert to byte array representation
    public toBytes(): Uint8Array {
        return requestIDToBytes(this);
    }

    // human-readable string representation
    public toString(): string {
        return requestIDToString(this);
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export function requestIDDecode(dec: WasmDecoder): ScRequestID {
    return requestIDFromBytesUnchecked(dec.fixedBytes(ScRequestIDLength));
}

export function requestIDEncode(enc: WasmEncoder, value: ScRequestID): void {
    enc.fixedBytes(value.id, ScRequestIDLength);
}

export function requestIDFromBytes(buf: Uint8Array | null): ScRequestID {
    if (buf === null || buf.length == 0) {
        return new ScRequestID();
    }
    if (buf.length != ScRequestIDLength) {
        panic("invalid RequestID length");
    }
    // final uint16 output index must be > ledgerstate.MaxOutputCount
    if (buf[ScRequestIDLength - 2] > 127 || buf[ScRequestIDLength - 1] != 0) {
        panic("invalid RequestID: output index > 127");
    }
    return requestIDFromBytesUnchecked(buf);
}

export function requestIDToBytes(value: ScRequestID): Uint8Array {
    return value.id;
}

export function requestIDFromString(value: string): ScRequestID {
    return requestIDFromBytes(hexDecode(value));
}

export function requestIDToString(value: ScRequestID): string {
    return hexEncode(requestIDToBytes(value));
}

function requestIDFromBytesUnchecked(buf: Uint8Array): ScRequestID {
    let o = new ScRequestID();
    o.id = buf.slice(0);
    return o;
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export class ScImmutableRequestID {
    proxy: Proxy;

    constructor(proxy: Proxy) {
        this.proxy = proxy;
    }

    exists(): bool {
        return this.proxy.exists();
    }

    toString(): string {
        return requestIDToString(this.value());
    }

    value(): ScRequestID {
        return requestIDFromBytes(this.proxy.get());
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export class ScMutableRequestID extends ScImmutableRequestID {
    delete(): void {
        this.proxy.delete();
    }

    setValue(value: ScRequestID): void {
        this.proxy.set(requestIDToBytes(value));
    }
}
